lib/files.js

```diff
@@ -0,0 +1,86 @@
+/**
+ * Create a file
+ *
+ * @param {object} file - Required. The file object.
+ * @param {object} metadata - Optional. User-provided metadata associated with the file.
+ * @returns {Promise<object>} - Resolves with the file data
+ */
+async function createFile(file, metadata = {}) {
+  const form = new FormData();
+
+  let filename;
+  let blob;
+  if (file instanceof Blob) {
+    fileName = file.name || `blob_${Date.now()}`;
+    blob = file;
+  } else if (Buffer.isBuffer(file)) {
+    fileName = `buffer_${Date.now()}`;
+    blob = new Blob(file, { type: "application/octet-stream" });
+  } else {
+    throw new Error("Invalid file argument, must be a Blob or File");
+  }
+
+  form.append("content", blob, filename);
+  form.append(
+    "metadata",
+    new Blob([JSON.stringify(metadata)], { type: "application/json" })
+  );
+
+  const response = await this.request("/files", {
+    method: "POST",
+    data: form,
+    headers: {
+      "Content-Type": "multipart/form-data",
+    },
+  });
```

Comment:
Absolutely. HTTP pedant I am, I'd been thinking to implement this as a conditional `POST` using `If-None-Match`. But I'm pessimistic about clients consistently sending the request header without also sending the form body (and thereby nullifying any benefit).

So instead, I'm thinking this could be an entirely client-side optimization. Get the first page of files at the top of `transformFileInputsToReplicateFileURLs`, and then check against the Etags / shasums in that. That would also have the benefit of allowing users to control exactly when to reuse existing files.